src/ddsrt/include/dds/ddsrt/misc.h

```diff
@@ -89,6 +89,12 @@ extern "C" {
   DDSRT_WARNING_GNUC_ON(deprecated-declarations) \
   DDSRT_WARNING_MSVC_ON(4996)
 
+/**
+ * @brief Declare main function to fix the no previous prototype warning
+ *  when main function is wrapped (e.g. in the FreeRTOS port)
+ */
+int main(int argc, char **argv);
+
```

Comment:
This seems very odd, I'm 99,99% sure this isn't what we want to do.

Reply:
I fully agree that this seems very odd, but I think it is correct (as `main` is replaced by something else with a preprocessor define and becomes an ordinary function). I'll add `#ifdef main` around the declaration, maybe that helps? I don't see any alternatives, except for leaving this warning for now (but that means we cannot enable `-Werror` on FreeRTOS builds)

Reply:
I think we shouldn't (re)define `main` ever, especially not in headers that are public. The compiler is likely complaining about `realmain`, which is a hack in itself, not being declared before it's defined(?) In that case providing a header that's always prepended (there's a `gcc` option for that) could be a fix. Another option could be to rename the symbol. In any case, it's definitely a problem in the FreeRTOS build, therefore the place to fix it is in that specific build.

Reply:
This code is not redeclaring `main`, but redeclaring the function with the rewritten name (`real_main` in this case), as the declaration is also rewritten by the define that is provided with the gcc `-D main=real_main` option. You're right that it should not be in a public header, it should be in a specific place for the FreeRTOS build. 

What I could do is add a header file like this:

```C
#ifndef FREERTOS_MAINWRAP_H
#define FREERTOS_MAINWRAP_H

/**
 * @brief Declare main function to avoid the 'no previous prototype' warning
 *  when main function is wrapped (e.g. in the FreeRTOS port). In that case the
 *  main function is renamed to real_main by using a definition for main passed
 *  to the compiler on the command line (-D main=real_main).
 */
#ifdef main
int main(int argc, char **argv);
#endif

#endif /* FREERTOS_MAINWRAP_H */
```
This header should be in a freertos specific location in the Cyclone source tree (any suggestions?), and that location needs to be in the include_path when running a FreeRTOS build. The header can then be prepended using the gcc `-include` option you mentioned. 

For now I've removed the declaration, so that we can finish this PR and fix the issue later in a separate PR.

Reply:
@dennis-adlink, the fix is absolutely correct, but it's the wrong repository :slightly_smiling_face: The BSP for this particular build contains a small wrapper that initializes the embedded device. There should be a CMake toolchain file there as well where you can add the `-include` option to specify that a header file of your choosing must be included first, that's probably the best place to add it.